File: SegundaParte/Semana8/Python/CG.py
class Renderable:
    def render(self):
        print("Renderizando objeto.")

class Sprite(Renderable):
    def animate(self):
        print("Animação do sprite.")

class Particle(Renderable):
    def update(self):
        print("Atualizando partícula.")

class GameObject(Sprite, Particle):
    def interact(self):
        print("Interagindo com o objeto.")

    # Método update diretamente em GameObject
    def update(self):
        print("Atualizando objeto do tipo GameObject.")

meu_objeto = GameObject()

meu_objeto.render()   # Herdado de Renderable
meu_objeto.animate()  # Herdado de Sprite
meu_objeto.update()   # Diretamente em GameObject
meu_objeto.interact()  # Próprio método de GameObject

File: SegundaParte/Semana8/Java/CG.java
interface Renderable {
    void render();
}

class Sprite implements Renderable {
    public void animate() {
        System.out.println("Animação do sprite.");
    }

    @Override
    public void render() {
        System.out.println("Renderizando objeto.");
    }
}

class Particle implements Renderable {
    public void update() {
        System.out.println("Atualizando partícula.");
    }

    @Override
    public void render() {
        System.out.println("Renderizando objeto.");
    }
}

class GameObject extends Sprite {
    public void interact() {
        System.out.println("Interagindo com o objeto.");
    }

    public void update() {
        System.out.println("Atualizando objeto do tipo GameObject.");
    }
}

public class CG {
    public static void main(String[] args) {
        GameObject meuObjeto = new GameObject();
        ((Renderable) meuObjeto).render(); // Herdado de Renderable
        meuObjeto.animate(); // Herdado de Sprite
        meuObjeto.update(); // Diretamente em GameObject
        meuObjeto.interact(); // Próprio método de GameObject
    }
}

